File: src/main/java/com/droiddevgeeks/crashsink/CrashInterceptor.java
package com.droiddevgeeks.crashsink;

/** Decorator in the uncaught-exception chain. Captures our crashes, always delegates. */
public final class CrashInterceptor implements Thread.UncaughtExceptionHandler {

    private final Thread.UncaughtExceptionHandler previous;
    private final CrashAttributor attributor;
    private final CrashProcessor processor;

    public CrashInterceptor(final Thread.UncaughtExceptionHandler previous,
                            final CrashAttributor attributor,
                            final CrashProcessor processor) {
        this.previous = previous;
        this.attributor = attributor;
        this.processor = processor;
    }

    public Thread.UncaughtExceptionHandler getPrevious() {
        return previous;
    }

    @Override
    public void uncaughtException(final Thread thread, final Throwable throwable) {
        try {
            if (attributor.isOurs(throwable)) {
                processor.persistBlocking(thread, throwable);
            }
        } catch (Throwable suppressed) {
            // Our reporting must never mask the original crash.
        } finally {
            if (previous != null) {
                previous.uncaughtException(thread, throwable);
            } else {
                System.exit(10);
            }
        }
    }
}

File: crashsink/src/main/kotlin/com/droiddevgeeks/crashsink/CrashInterceptor.kt
package com.droiddevgeeks.crashsink

/** Decorator in the uncaught-exception chain. Captures our crashes, always delegates. */
class CrashInterceptor(
    val previous: Thread.UncaughtExceptionHandler?,
    val attributor: CrashAttributor,
    private val processor: CrashProcessor
) : Thread.UncaughtExceptionHandler {

    override fun uncaughtException(thread: Thread, throwable: Throwable) {
        try {
            if (attributor.isOurs(throwable)) {
                processor.persistBlocking(thread, throwable)
            }
        } catch (suppressed: Throwable) {
            // Our reporting must never mask the original crash.
        } finally {
            val prev = previous
            if (prev != null) {
                prev.uncaughtException(thread, throwable)
            } else {
                System.exit(10)
            }
        }
    }
}
